[WIP] drivers: CAN: Extend the CAN API to support CAN-FD

alexanderwachter opened the pull request:
This PR is a proposal for making the CAN API to support CAN-FD.
The stm32fd driver is currently empty and serves dummy implementation for testing device-tree generation.

alexanderwachter committed:
include: drivers: can: rework zcan_frame and zcan_filter

This commit is a rework of the zcan_frame and zcan_filter
structs. The packed attribute is removed, and the data
can be aligned according to the needs of the CPU.
The size-penalty for the zcan_frame is two bytes;
the zcan_filter is reduced by the half.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
tests: isotp: Make sure sending is finished after a test.

This PR adds a semaphore to make sure that the send function
callback is called before the next testcase starts.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter commented:
ping @martinjaeger

mickdermack commented on include/drivers/can.h:
```diff
@@ -184,8 +188,8 @@ struct zcan_frame {
-		u8_t data[8];
-		u32_t data_32[2];
+		u8_t data[CAN_MAX_DLEN];
+		u32_t data_32[CAN_MAX_DLEN / sizeof(u32_t)];
```
Calculating the size like this does not allow access to the whole data via `data_32` if `CAN_MAX_DLEN` is not a multiple of `sizeof(u32_t)` i.e. 4 bytes. Might make sense to round up here.

karstenkoenig commented on include/drivers/can.h:
```diff
@@ -325,14 +334,95 @@ struct can_driver_api {
+ * @retval CAN_TX_* on failure.
+ */
+__syscall int can_fd_send(struct device *dev, const struct zcan_frame *frame,
+		       bool brs, bool fd_format, s32_t timeout,
```
Unsure if there is a guideline about this, but this is becoming quite a long list of parameters. Maybe the bools can be merged into a flag bitmask?

alexanderwachter replied:
Maybe we can add this flags to the zcan_frame struct. I‘ll check that next week. Maybe we have some space thats padding so far.

alexanderwachter replied:
With the changes from https://github.com/alexanderwachter/zephyr/blob/can_rework_structs/include/drivers/can.h we could use the first reserved bit to distinguish fd frames.

karstenkoenig commented on drivers/can/Kconfig:
```diff
@@ -32,6 +32,22 @@ config CAN_INIT_PRIORITY
+	default 64 if CAN_FD_MODE
+	range 8 64
+	help
+	  Defines the data length for CAN frames.
```
Maybe say maximum data length for CAN frames

henrikbrixandersen replied:
Should this be user-selectable for non-CAN-FD? I think it should be hardcoded to 8 for non-CAN-FD.

alexanderwachter replied:
It is set to eight automatically when we have classic CAN. And the range does not allow to have less than eight.

henrikbrixandersen replied:
What I meant was that we should not expose this Kconfig setting to the user if CAN-FD is not enabled.

karstenkoenig reviewed (approved):
Looks like a sane approach overall :-)

mickdermack commented on drivers/can/Kconfig.stm32fd:
```diff
@@ -0,0 +1,32 @@
+	depends on SOC_SERIES_STM32G4X
+	help
+	  Enable STM32 CAN-FD Driver.
+	  This diver supports the FDCAN IP from STM32G4 SoCs
```
```suggestion
	  This driver supports the FDCAN IP of STM32G4 SoCs
```

alexanderwachter committed:
drivers: can: Extend the CAN API to be compatible to CAN-FD

This commit makes the CAN API compatible to CAN-FD.
To do so, CAN_MAX_DLEN is set by Kconfig to up to 64 bytes.
Two flags are introduced in the zcan frames.
One is the fd-flag which signals a CAN-FD frame and the
other is the brs to signal a bitrate switch.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
drivers: can: Implement Bosch M_CAN driver

Implementation of the Bosch M_CAN IP driver.
This driver is just the base of a specific SoC implementation.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter commented:
stm32fd driver passes classic CAN API tests now. CAN-FD still needs to be tested.

alexanderwachter committed:
tests: driver: can: api: Enhanced tests to check dispatching

This commit enhances the CAN API test. In the send_receive tests,
we are using two filters and frames at the same time to check
if the frame gets dispatched to the correct filter.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
canbus: iso-tp: Use the dedicated workq.

The implementation used the system workq instead of the
dedicated workq. Use k_work_submit_to_queue() instead of
k_work_submit() now.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
drivers: can: Rework can_configure API

The previous API can't change the sampling-point and only allowed
bitrates that fit the time segments.
The new API allows for shifting the sampling-point and adjusts the
number of time quantum in a bit to all more possible bitrates.
The functions to calculate the timings are moved to the can_common file.
They can be used for all drivers.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
drivers: can: SAM MCAN Implementation

Specialisation of the Bosch M_CAN driver for the ATMEL
SAM SoCs.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
drivers: can: Implement stm32fd driver

This driver is the SoC specific implementation of the
Bosch M_CAN IP.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

nandojve commented on drivers/can/can_sam.c:
```diff
@@ -0,0 +1,239 @@
+
+static const struct can_sam_config can_sam_cfg_0 = {
+	.pmc_id = DT_PROP(DT_NODELABEL(can0), peripheral_id),
+	/*.pin_list = {
+		ATMEL_SAM_DT_PIN(0, 0),
+		ATMEL_SAM_DT_PIN(0, 1)
+	},*/
```
pin list need listfy, you need use:
`.pin_list = ATMEL_SAM_DT_PINS(n),`
You can need to:
`.num_pins = ATMEL_SAM_DT_NUM_PINS(n),`
Take look at drivers/spi/spi_sam.c
The definition is here:
./soc/arm/atmel_sam/common/atmel_sam_dt.h
The idea is all Atmel DT macros should be place there to easy reuse.

nandojve commented on drivers/can/can_sam.c:
```diff
@@ -0,0 +1,239 @@
+		    DT_IRQ_BY_NAME(DT_NODELABEL(can0), line_1, priority),
+		    can_sam_line_1_isr, DEVICE_GET(can_mcan_0), 0);
+	irq_enable(DT_IRQ_BY_NAME(DT_NODELABEL(can0), line_1, irq));
+}
```
This drive should be must instance because we have can0 and can1.
Try follow drivers/spi/spi_sam.c and add
`DT_INST_FOREACH_STATUS_OKAY(SPI_SAM_DEVICE_INIT)`

nandojve commented on dts/arm/atmel/same70-pinctrl.dtsi:
```diff
@@ -108,6 +108,10 @@
 			DT_ATMEL_PIN(usart2, rxd2, d, 15, b);
 			DT_ATMEL_PIN(usart2, sck2, d, 17, b);
 			DT_ATMEL_PIN(usart2, txd2, d, 16, b);
+			DT_ATMEL_PIN(mcan0, canrx0, b, 16, a);
+			DT_ATMEL_PIN(mcan0, cantx0, b, 2, a);
+			DT_ATMEL_PIN(mcan1, canrx1, c, 12, c);
+			DT_ATMEL_PIN(mcan1, cantx1, c, 14, c);
```
From Atmel SAMV71 datasheet:
![atmel_mcan_pins](https://user-images.githubusercontent.com/7027887/82150212-752ca880-982e-11ea-9b15-cf4e61ee8058.png)

nandojve commented on dts/arm/atmel/same70.dtsi:
```diff
@@ -381,6 +381,32 @@
+				label = "CAN_0";
+				peripheral-id = <35>;
+				pinctrl-0 = <&pb16a_mcan0_canrx0 &pb2a_mcan0_cantx0>;
+			};
```
Any particular reason to not add can1 instance?

alexanderwachter replied:
I'll add it when CAN0 works :-)

nandojve commented on soc/arm/atmel_sam/same70/Kconfig.defconfig.series:
```diff
@@ -84,4 +84,8 @@ config PWM_SAM
 	default y
 	depends on PWM
 
+config CAN_SAM
+	default y
+	depends on CAN
+
```
Please, add this to samv71 too.

alexanderwachter replied:
I'll try to use the "new way" to do it. See STM32 implementation

nandojve reviewed (changes requested):
Nice work! I hope to see this soon on master :D

nandojve commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -298,7 +298,7 @@
 				#address-cells = <1>;
 				#size-cells = <0>;
-				reg = <0x40006400 0x400>, <0x4000B400 0x350>;
+				reg = <0x40006400 0x400>, <0x4000A400 0x350>;
```
This patch is for STM32, not SAM SoC

martinjaeger commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -283,6 +283,54 @@
+				label = "CAN_1";
+			};
+
+			can2: can@40006800 {
+				compatible = "st,stm32-fdcan";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0x40006800 0x400>, <0x4000A800 0x350>;
+				reg-names = "m_can", "message_ram";
+				interrupts = <86 0>, <87 0>;
+				interrupt-names = "LINE_0", "LINE_1";
+				status = "disabled";
+				label = "CAN_2";
+			};
+
+			can3: can@40006C00 {
+				compatible = "st,stm32-fdcan";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0x40006C00 0x400>, <0x4000AC00 0x350>;
+				reg-names = "m_can", "message_ram";
+				interrupts = <88 0>, <89 0>;
+				interrupt-names = "LINE_0", "LINE_1";
+				status = "disabled";
+				label = "CAN_3";
+			};
```
FDCAN2 and FDCAN3 are only available in STM32G474, so these nodes should be deleted for STM32G431.

martinjaeger commented on drivers/can/Kconfig.stm32fd:
```diff
@@ -0,0 +1,40 @@
+
+DT_COMPAT_STM32_FDCAN := st,stm32-fdcan
+
+config CAN_STM32FD
+	bool
+	default $(dt_compat_enabled,$(DT_COMPAT_STM32_FDCAN))
```
It makes sense to automatically enable the FD driver based on the devicetree node. Ideally, the CAN driver `CAN_STM32` for older MCUs would be disabled automatically aswell (in `Kconfig.stm32`) as it still shows up in menuconfig and causes compile errors if enabled.

martinjaeger reviewed (commented):
Currently testing this out on an STM32G431.

Did I understand correctly, that also ST uses the Bosch M_CAN IP? Couldn't find this anywhere in the reference manual.

martinjaeger commented:
@alexanderwachter I know that the SAM drivers don't work well yet, but the one for STM32 works like a charm. Do you think you could split this PR so that we can get the STM32 part merged already (maybe even for 2.4)?

alexanderwachter commented:
Let me try. There is also a lot of work for porting the existing drivers to the new timing API.

martinjaeger commented:
> Let me try. There is also a lot of work for porting the existing drivers to the new timing API.

Ok, I was not aware that there is much work left also for the STM32 drivers. Don't worry too much, Zephyr v2.5 is probably more realistic then. I will continue with testing and let you know if I encounter any issues with the driver.

y2sh commented:
hi @alexanderwachter , I am currently working on a custom board with Atmel V71...b , and I would like to check its canbus over zephyr. I see that you are working on it, and I can help test it on my board.

alexanderwachter commented:
Hi @y2sh. This PR is functional on stm32 only at the moment. I tried it on the SAM70Explaned, but it did not work.
I think the problem is somewhere in the msg SRAM. I would be happy if someone with SAM experience could help me to spot the problem. As @martinjaeger already mentioned, this PR needs some rework. The first thing that needs to go in is #28345 .
After that, we need to decide how we can optimize the zcan_frame struct.

y2sh commented:
I am busy with other issues, so no development from my side for now. However, I can help with testing canbus relating code on my board. I just saw that #28345 didn't pass the pull request, once it will I can run some canbus tests on my board.
In the meanwhile I am attaching the mcan config file of Atmel studio, from my experience I can tell it is work great.
(I changed it from .h to .txt , I cannot attach code files)
[conf_mcan.txt](https://github.com/zephyrproject-rtos/zephyr/files/5375768/conf_mcan.txt)

y2sh commented:
Hi @alexanderwachter ,  I started to write code of canbus , currently I am relying on the implementation of the Atmel ASF functions, but I am missing some basic understanding of the zephyr infra-structure - could you help me understand at which point the "include" of the board is done? I meaning to mcan.h , which is in my pc locate at  "C:\Users\Administrator\zephyrproject\modules\hal\atmel\asf\sam\include\samv71b\component\mcan.h" ? 10x , Yair.

alexanderwachter commented:
@y2sh what would be a question for the general slack channel.
Usually, those things are included in soc.h

y2sh commented:
Hi @alexanderwachter , GREAT NEWS !!
I made the canbus work in Atmel V71b !!
still there is a bug (I am trying to solve it right now) , but the very basic of can send is working !
I did it "fast and *very* dirty", apologize for that. also as I wrote earlier I am missing some understanding in 
Zephyr infrastructure,  so I changed the code under "can_stm32.c / .h"
I am attaching the code, if you have interest I will write in details the bug.
Still, currently I implement only can_send, without isr and callbacks, but it work.
cheers, Yair.
[New folder.zip](https://github.com/zephyrproject-rtos/zephyr/files/5409029/New.folder.zip)

alexanderwachter commented:
Implemented by #28345 and follow-up with stm32 fdcan driver.
